numpy/core/src/multiarray/getset.c

```diff
@@ -477,41 +477,99 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
         }
     }
 
-
-    if ((newtype->elsize != PyArray_DESCR(self)->elsize) &&
-        (PyArray_NDIM(self) == 0 || !PyArray_ISONESEGMENT(self) ||
-         PyDataType_HASSUBARRAY(newtype))) {
-        goto fail;
-    }
-    if (PyArray_ISCONTIGUOUS(self)) {
-        i = PyArray_NDIM(self) - 1;
-    }
-    else {
-        i = 0;
-    }
-    if (newtype->elsize < PyArray_DESCR(self)->elsize) {
+    if (new_elsize != old_elsize) {
         /*
-         * if it is compatible increase the size of the
-         * dimension at end (or at the front for NPY_ARRAY_F_CONTIGUOUS)
+         * One can get creative as to when a view of an array with a
+         * dtype of different item size is possible. The following
+         * rules are pretty flexible, and still guarantee that the view
+         * can be undone, i.e. that viewing the new view with the
+         * original dtype returns the original array:
+         *
+         *  1. The array must be contiguous along the dimension with
+         *     the smallest absolute value stride, i.e. the stride's
+         *     absolute value must be equal to the item size.
+         *  2. There should be only one dimension having this
+         *     minimal stride.
+         *  3. The new item size must evenly divide the total size of
+         *     the minimal stride dimension.
+         *
+         * Conditions 1 and 3 could be relaxed, e.g. to allow
+         * overlapping arrays, i.e. with a stride smaller than the item
+         * size, of to discard extra bytes from the view, but this
+         * would not allow the view to be undone.
+         * Condition 2 is required to avoid indexing beyond the limits
+         * of the array's allocated memory.
          */
-        if (PyArray_DESCR(self)->elsize % newtype->elsize != 0) {
+        npy_intp stride, axis_size;
+        int j, axis;
+        int nd = PyArray_NDIM(self);
+
+        if (nd == 0 || PyDataType_HASSUBARRAY(newtype)) {
```

Comment:
That's definitely out of the scope of this PR, but yes, it is a good question for which I don't have a definite answer. It would seem that, as long as there are no Python objects, all you should care about is the element size. Actually, with this PR you can do the following, which removes the `[0]` (but would work with a 1 item slice like `[:1]`) and uses an intermediate conversion to a void dtype:

```
>>> y['head'].view('|V12').view('>u4')[:] = 5
>>> y
array([((5L, 5L, 5L), [0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L])],
      dtype=[('head', [('dx', '>u4'), ('dy', '>u4'), ('dz', '>u4')]), ('data', '>u4', (10,))])
```

So unless I am missing something horribly bad that could be happening, that check should also be revisited.